### DISCUSSION TITLE
[Draft] Relative path handling in Astro

### DISCUSSION THREAD

**Discussion by jonathantneal** [withastro/roadmap#23] [Proposal]
<h1 align="center">Relative path handling in Astro</h1>

Astro should make it easier to reference images and assets relative to the source file.

<details><summary id="md-example"><strong>Examples</strong>:</summary>

```astro
<img
  alt="Clock"
  local:src="./my-kitten.avif"
  local:srcset="
    ./my-kitten.avif 1x,
    ./my-kitten@2x.avif
  "
/>
```

```astro
---
import Kitten1x from './my-kitten.avif' assert { type: "url" }
import Kitten2x from './my-kitten@2x.avif' assert { type: "url" }
import linkToScript from './script.js' assert { type: "url" }
---
```

</details>

## Background

In Astro HTML, it’s not always easy to know where a **relative** path points to.

```astro
<img src="./kitten.avif" alt="kitty" />
```

If a relative path is used inside of a component, then it might appear to be a reference to something relative to that component _in the source_. Instead, the path will be relative to the URL where the component was used _in the final output_.

### Goals

- Provide an intuitive way to reference locations relative to the source file.
- Provide an intuitive way to reference early (in Frontmatter) and late (in HTML).

### Non Goals

- Certain implementation details like API naming, argument shape, etc., because that can be decided later.
- Issues with relative 'links' that start with `/`, because it’s much more complicated.
  - A link starting with `/` is a relative link, while a path starting with `/` is an absolute path.
  - Vite 'muddies' this distinction by allowing `/` to be used as a relative link, or as an absolute path, or as a project relative path.
  - This RFC focuses on relative links that are also relative paths and calls them 'paths', for simplicity.
- An image component, because it’s much more complicated.
  - This RFC just happens to use a lot of image examples for simplicity.
  - Relative paths also affect `<a>`, `<link>`, `<script>`, etc. and Astro users continue to face issues with these [<sup>1</sup>](https://discord.com/channels/830184174198718474/845451724738265138/909188442653610024) [<sup>2</sup>](https://discord.com/channels/830184174198718474/845451724738265138/909468145075494944) [<sup>3</sup>](https://discord.com/channels/830184174198718474/909920313284890654/909951130795143249).
  - An image component could be a followup to this, focusing on other image-related issues like transformation & compression.

---

## Current Solutions:

Astro currently supports two solutions; one for 'late' resolution in the HTML block and one for 'early' resolution in the Frontmatter block.

### Relative URLs in `Astro.resolve`

> Current Solution 1 for late resolution

To author a relative path that is relative _to the source_, use `Astro.resolve` in Astro HTML.

```astro
<!-- src/components/my-kitten.astro -->
<a href={Astro.resolve('./my-kitten.avif')}>
  <img src={Astro.resolve('./my-kitten.avif')} />
</a>
```

**Pros**:

* This generates a path relative to the source that remains true in the final output.
* The path stays written alongside its context.
* This does not break the way JavaScript works.

**Cons**:

* The author must introduce a JavaScript block into their Astro HTML.
* The JS block interupts the flow of well-formed HTML.
  * This is even more problematic when `srcset` is used:
    ```astro
    <img
      alt="Clock"
      src={Astro.resolve('./my-kitten.avif')}
      srcset=`
        ${Astro.resolve('./my-kitten.avif')} 1x,
        ${Astro.resolve('./my-kitten@2x.avif')}
      `
    />
    ```

### Relative URLs in JavaScript Imports

> Current Solution 2 for early resolution

To author a relative path that is relative _to the source_, use an `import` in Astro Frontmatter.

```astro
---
import Kitten1x from './my-kitten.avif'
import Kitten2x from './my-kitten@2x.avif'
---
<img
  alt="Clock"
  src={Kitten1x}
  srcset=`
    ${Kitten1x} 1x,
    ${Kitten1x}
  `
/>
```

**Pros**:

* This generates a path relative to the source that remains true in the final output.
* The JS block interupts the flow of well-formed HTML much less than `Astro.resolve`.

**Cons**:

* The author must introduce a Frontmatter block into their file and a JavaScript block into their Astro HTML.
  * In cases like `srcset`, this still lacks cohesiveness with HTML.
* The author must move their URL away from its context.
* The author must give a meaningful variable name to the URL.
* This breaks the way JavaScript works.
* This prevents certain kinds if relative paths to be imported.
  * This is even more problematic if the asset _might_ also be import-able to JS:
    ```astro
    ---
    import linkToScript from './script.js'
    ---
    ```

---

## Proposed Solutions

Astro should continue to support two solutions; one for late resolution in the HTML block and one for early resolution in the Frontmatter block.

### Relative URLs in Namespaced Attributes

> Proposed Solution 1 for late resolution

To author a relative path that is relative _to the source_, use a `local` namespaced attribute.

```astro
<img
  alt="Clock"
  local:src="./my-kitten.avif"
  local:srcset="
    ./my-kitten.avif 1x,
    ./my-kitten@2x.avif
  "
/>
```

The `local` namespace would resolve relative paths in the attribute value while dropping the namespace from the attribute in the final output.

**Pros**:

* This generates a path relative to the source that remains true in the final output.
* This source is very cohesive with HTML.
* The author does not need to escape into JavaScript.
  * Which means this also does not break the way JavaScript works.
  * Which means the author does not also need to name any JavaScript variable.

**Cons**:

* The `local` namespace would need to be parsed to locate and transform relative paths.
* The 'late' processsing may increase the time it takes Astro to detect relative paths and add them to an asset graph.

## Relative URLs in Asserted JavaScript Imports

> Proposed solution 2 for early resolution

To author a relative path that is relative _to the source_, use an `import` with assertions in Astro Frontmatter.

```astro
---
import Kitten1x from './my-kitten.avif' assert { type: "url" }
import Kitten2x from './my-kitten@2x.avif' assert { type: "url" }
import linkToScript from './script.js' assert { type: "url" }
---
```

These type assertions would communicate how the file should be interpretted by JavaScript.

**Pros**:

* This generates a path relative to the source that remains true in the final output.
* This respects the way JavaScript works.
* This allows any relative paths to be imported.
* This syntax is permitted by our VSCode extension.

**Cons**:

* The author must introduce a Frontmatter block into their file and a JavaScript block into their Astro HTML.
  * In cases like `srcset`, this still lacks cohesiveness with HTML.
* The author must move their URL away from its context.
* The author must give a meaningful variable name to the URL.
* This syntax is not yet supported by the Astro compiler.

---

### Relative URLs with a Custom Scheme

> Alternative solution 1 for late resolution

To author a relative path that is relative _to the source_, use a `local` schema.

```astro
<img
  alt="Clock"
  src="local:./my-kitten.avif"
  srcset="
    local:./my-kitten.avif 1x,
    local:./my-kitten@2x.avif
  "
/>
```

**Pros**:

* This generates a path relative to the source that remains true in the final output.
* This source is very cohesive with HTML.
* The author does not need to escape into JavaScript.
  * Which means this also does not break the way JavaScript works.
  * Which means the author does not also need to name any JavaScript variable.
* The schema could be supported in other APIs (like JavaScript URLs) for consistency.

**Cons**:

* The `local` schema would need to be parsed in every attribute to locate and transform relative paths.
* The 'late' processsing may increase the time it takes Astro to detect relative paths and add them to an asset graph.

### Relative URLs with Query Parameters

> Alternative solution 2 for early resolution

To author a relative path that is relative _to the source_, use an `import` with query parameters in Astro Frontmatter.

```astro
---
import Kitten1x from './my-kitten.avif?type=url'
import Kitten2x from './my-kitten@2x.avif?type=url'
---
```

These query parameters communicate how the file should be loaded.

**Pros**:

* This generates a path relative to the source that remains true in the final output.
* This allows any relative paths to be imported.
* This syntax is permitted by our VSCode extension, and Vite may already support it.

**Cons**:

* The author must introduce a Frontmatter block into their file and a JavaScript block into their Astro HTML.
  * In cases like `srcset`, this still lacks cohesiveness with HTML.
* The author must move their URL away from its context.
* The author must give a meaningful variable name to the URL.
* This breaks the way JavaScript works (without `assert { type: 'json' }`).

---

### Should all relative paths without custom syntax warn or throw? 

No, Astro HTML should respect well known and well established conventions in HTML. If you intentionally use relative paths and are comfortable with this context, you’ll be very glad to know that Astro won’t make you change valid HTML.

### Should Astro Support Multiple Solutions?

Yes, Astro needs to support late resolution and early resolution. For late resolution in Astro HTML, the solution should be cohesive with HTML. For early resolution in Astro Frontmatter, the solution should be cohesive with JavaScript.

### How would import assertions work in Astro?

1. The compiler handling Astro Frontmatter would need to pass thru import assertions.
2. The Acorn plugin for import assertions would need to added to Rollup.
3. The Rollup transform handler would need to return the resolved id path when the assertion instructed so.

**Comment by FredKSchott** [comment]
> Yes, Astro needs to support late resolution and early resolution. For late resolution in Astro HTML, the solution should be cohesive with HTML.

In writing https://github.com/withastro/rfcs/discussions/28 I realized that `Astro.resolve()` and the concept of "late resolution" may be entirely incompatible with SSR. I'll leave it to @matthewp to give any final say here, but I can't think of a way that we could get around this issue:

CSS and JS are bundled based on the known import graph. If `Astro.resolve()` runs at runtime, it could reference a CSS or JS file that doesn't actually exist in the final build because it wasn't detected as a file entrypoint, and was instead bundled into a larger bundle of CSS or JS. I can't think of a way around this that is compatible with the idea of bundling.

The promise of `src` vs. `public` is that we own `src` completely, and have full ability to bundle and optimize as needed. We make no promises that files in `src` will exist 1:1 in the final output. You can see this clearly defined here: https://docs.astro.build/core-concepts/project-structure/#src

I mainly want to flag this because you phrased this as an absolute, that Astro needs to support late runtime resolution, but I don't think that this is true, at least for v1.0. Plenty of frameworks get around this limitation today without too much pain, as far as I can tell. This may change your proposal if this ends up to be the case.

**Reply by FredKSchott** [reply]
Update: Also note that Vite uses this plugin internally to support some decent static analysis when used with dynamic ESM imports: https://github.com/rollup/plugins/tree/master/packages/dynamic-import-vars. If we use Vite to drive our build, then we could use static ESM imports with the `?url` flag for most assets + dynamic ESM `import()` with the `?url` flag for late resolution as an advanced use-case. 

If we wanted to make late resolution easier, then maybe we could compile `Astro.resolve('...')` to `await import('...?url')` automatically for the user as some nice sugar, similar to how `Astro.fetchContent` is rewritten to use `import.meta.globEager` automatically for the user.

**Reply by jonathantneal** [reply]
Maybe this part of the proposal lacks clarity, or it’s somewhat out of scope.

If the author uses a static value, like `<img local:src="./relative.jpg" />`, then it **can** be analyzed.

If the author uses a dynamic value, like `<img local:src={dynamic} />`, then it **cannot** be analyzed.

I phrased the need for 'late' resolution as an absolute, but only because I think a 'late' API (like _`local:src` or whatever it ends up being_) is critical to the authoring experience. This is a key observation made in the **Pros** and **Cons**.

Whether 'early' or 'late' resolution can be analyzed or not seems like a separate, albeit important issue.

**Reply by FredKSchott** [reply]
So to be sure I'm understanding, you're proposing that the compiler would error if it saw anything other than a string literal in `local:src`? In that case, is it fair to consider this more like syntactic sugar vs. an ESM import with `?url`? Both accomplish mostly the same feature set with the same drawbacks, if I understand the pros/cons correctly. I guess I'm having trouble squaring this with the absolute language around why late resolution is "critical", "needs to be supported", etc.

```astro
---
import imgUrl from '../images/foo.png';
---
<img src={imgUrl} />
```

```astro
<img local:src='../images/foo.png' />
```
If we consider this sugar, then I'm definitely not opposed post-v1.0! But I think a v1.0 release would take priority over this given that it adds no new features and is fairly subjective over which the user would prefer.

**Reply by FredKSchott** [reply]
Being more specific to the part of the RFC that is relevant, I am saying that I believe that this line in the RFC:

> "The 'late' processsing may increase the time it takes Astro to detect relative paths and add them to an asset graph."

Is a non-starter for SSR support. All asset graph detection must be early, if that asset lives in `src`, because otherwise we give no guarantees that it will be included in the final build. Or, the asset must live in public, in which case no resolution is needed and the asset can be referenced by absolute URL (`<img src="/foo.png"`). 

FWIW importing images from `public` is essentially how Next.js (and many SvelteKit blog posts) do it, so I'm starting to think this is more viable than we've given credit in past discussions!

**Reply by jonathantneal** [reply]
> So to be sure I'm understanding, you're proposing that the compiler would error

Have some time to discuss this in chat or a call? I’m not proposing that in this RFC. It’s not in the scope of this RFC.

This RFC is about how authors write relative references. It posits that authors reference things 2 ways — 'up top' or 'inline' — so to speak. It posits that 'inline' relative links are _essential_ because it’s contextual, colocated, close to how it works by default. That doesn’t mean anyone needs to agree, but it is the conclusion I’m running with.

If it helps to talk about it another way. Consider what Astro should do with this in v1.0:

```astro
---
const imgUrl = await import('../images/foo.png')
---
```

**Reply by jonathantneal** [reply]
> Is a non-starter for SSR support

Maybe I should not have written it the way I did. 😓

It’s 'late' in that the compiler needs to analyze HTML attributes.

But if that’s still a 'non-starter' then that seems harsh. If analyzing HTML attributes are a 'non-starter', then I would think that meant analyzing any attribute in HTML is a non-starter. There’d be no support for `<script src>`, `<link href>`, which are equivalent uses.

**Reply by FredKSchott** [reply]
Discussed in person, taking some notes so that we don't lose them!

- "late" should be defined, I had assumed it meant "runtime" but I now understand @jonathantneal meant something more like "scanned out of the component template and not the frontmatter"
- This RFC proposal is really doing two things: 1. Describing what we need to do about the fact that `Astro.resolve()` will no longer work in a post-SSR world (critical for v1.0), and 2. Sugar for a new `local:src` syntax that would reduce the boilerplate of using ESM imports for assets (nice to have). It will be much more productive to split this into two different RFCs.
- RE ESM imports + `?url` support: It would be helpful to make it explicitly clear that this RFC is aware that import assertions are not yet supported in Vite, and that `?url` is not an alternative solution but is actually a short-term solution that would naturally be replaced by import assertions once they're ready. This is an important part of the adoption plan (a new section in our RFC template!) and also will save us from having to revisit this all over again once import assertions are landed in Vite.
- RE The sugar `local:src` attribute: missing a summary of how it would be implemented, which is really important for deciding if this should be a part of v1.0 or not. Discussed this a bit, and the fact that `local:src` could _compile_ to a static ESM import in the compiler output, and then require no other changes across the rest of our codebase. This would gel nicely with our larger push to unify Astro behavior and not offer multiple APIs multiple different implementations.
- RE The sugar `local:src` attribute for images: Since we don't have an image component yet, it's not out of the question to just say that images must live in `public/` and be referenced by absolute path/URL (not resolved by Astro). This has its drawbacks but would solve the "how do dynamic run-time asset references" work. This is what Next.js does, and would be more than viable for v1.0 imo. Most importantly, it would buy us time to come up with the proper solution to solve with our own image component.

**Reply by jonathantneal** [reply]
Thanks for the great writeup. Agreed on all parts. One new thing I learned, tho –

> It would be helpful to make it explicitly clear that this RFC is aware that import assertions are not yet supported in Vite

After we chatted, I saw that Vite is receptive to these, but [maybe expecting us to do the work](https://github.com/vitejs/vite/issues/4934#issuecomment-925693717). Not sure if that factors in positively or negatively. 😅

**Reply by matthewp** [reply]
One issue with `import imgUrl from './image.png';` is that it will give you an absolute path like `/src/component/image.png`. This works fine in dev but won't for supporting subpath deployments. In 0.21 we update URLs to be relative in the final output, but we don't want to do that any more in 1.0. This might just be a Vite setting to append the appropriate subpath though.

**Reply by FredKSchott** [reply]
I think this is a Vite setting! at least, I think our subpath-during-dev behavior was inspired by them

**Comment by matthewp** [comment]
@jonathantneal I was confused by this discussion's title:

> Markdown Route Generation Proposal

Is this a mistake in copy/pasting? I think this is unrelated to the markdown idea, no?

**Reply by jonathantneal** [reply]
Good catch. Fixed!

**Comment by matthewp** [comment]
I think the `local:` sugar (assuming it is implemented as adding an import statement) would both be an easy thing to implement and would go along way with selling the removal of Astro.resolve.

**Comment by matthewp** [comment]
I think we should turn this into an RFC and include deprecating Astro.resolve as part of it.

**Comment by tohagan** [comment]
In case anyone should pass by: `Astro,resolved()` was removed in Astro v2.

- https://docs.astro.build/en/guides/upgrade-to/v2/#removed-astroresolve
